### ISSUE TITLE
ConfidenceThreshold 0 does not work together with RangeMatcher

### ISSUE THREAD

**Issue by Kalo0m** [lukasalexanderweber/stitching#56]
Hi,

I would like to use your library in order to make a 360 picture of a room (with a set of 12-13 pictures). I'm using band-width=1 because I know the order, confidence_threshold 0 and Subsetter(0). But when I try to link images, some links are a bit weird because features are only located in one place (as you can see below). Am I missing an option which can ce useful in my usecase? or maybe it's my camera which optimize settings (according to the environment) and create mismatches between images

![image](https://user-images.githubusercontent.com/61421713/223393879-ca790939-5bd2-4202-8b21-0a7f0a40785d.png)

Thanks for your help

**Comment by lukasalexanderweber** [comment]
Can you show the result?

**Comment by lukasalexanderweber** [comment]
can you give me the matches_graph_dot_file e.g. with `--matches_graph_dot_file matches.txt`

**Comment by Kalo0m** [comment]
```
graph matches_graph{
"test1.jpg" -- "test2.jpg"[label="Nm=64, Ni=31, C=1.13971"];
"test1.jpg" -- "test3.jpg"[label="Nm=0, Ni=0, C=0"];
"test1.jpg" -- "test4.jpg"[label="Nm=0, Ni=0, C=0"];
}
```
and the order is test1 - 2 - 3 - 4

but as you can see, there are some mismatch in the cabinet and on the radiator. Can we fix these?

**Comment by lukasalexanderweber** [comment]
thats odd. The matches graph should look like this:

graph matches_graph{
"test1.jpg" -- "test2.jpg"[label="Nm=64, Ni=31, C=1.13971"];
"test**2**.jpg" -- "test3.jpg"[label="Nm=0, Ni=0, C=0"]; -> not 0
"test**3**.jpg" -- "test4.jpg"[label="Nm=0, Ni=0, C=0"]; -> not 0
}

maybe there is a bug in `range_width`

**Comment by Kalo0m** [comment]
@lukasalexanderweber I've done some research in your code and it looks like the`cv.detail.matchesGraphAsString` method return always test1.jpg when we pass confidence_threshold=0. I've reproduced an example:

```python
import stitching

settings = {
    "range_width": 1,
    "confidence_threshold": 0.1,
    "matches_graph_dot_file": "matches_graph.dot"
}   
settings2 = {
    "range_width": 1,
    "confidence_threshold": 0,
    "matches_graph_dot_file": "matches_graph2.dot"
}   
stitcher = stitching.Stitcher(**settings)
stitcher2 = stitching.Stitcher(**settings2)

images = ["test1.JPG", "test2.JPG", "test3.JPG", "test4.JPG", "test5.JPG", "test6.JPG", "test7.JPG"]
stitcher.initialize_registration(images)
stitcher2.initialize_registration(images)

imgs = stitcher.resize_medium_resolution()
features = stitcher.find_features(imgs)
matches = stitcher.match_features(features)

imgs2 = stitcher2.resize_medium_resolution()
features2 = stitcher2.find_features(imgs2)
matches2 = stitcher2.match_features(features2)

stitcher.subset(imgs, features, matches) # works
stitcher2.subset(imgs, features, matches) # doesn't work
```
matches_graph.dot
```
graph matches_graph{
"test1.JPG" -- "test2.JPG"[label="Nm=62, Ni=31, C=1.16541"];
"test2.JPG" -- "test3.JPG"[label="Nm=66, Ni=31, C=1.11511"];
"test3.JPG" -- "test4.JPG"[label="Nm=54, Ni=39, C=1.61157"];
"test4.JPG" -- "test5.JPG"[label="Nm=30, Ni=17, C=1"];
"test5.JPG" -- "test6.JPG"[label="Nm=85, Ni=56, C=1.67164"];
"test6.JPG" -- "test7.JPG"[label="Nm=107, Ni=80, C=1.99501"];
}
```
matches_graph2.dot
```
graph matches_graph{
"test1.JPG" -- "test2.JPG"[label="Nm=64, Ni=31, C=1.13971"];
"test1.JPG" -- "test3.JPG"[label="Nm=0, Ni=0, C=0"];
"test1.JPG" -- "test4.JPG"[label="Nm=0, Ni=0, C=0"];
"test1.JPG" -- "test5.JPG"[label="Nm=0, Ni=0, C=0"];
"test1.JPG" -- "test6.JPG"[label="Nm=0, Ni=0, C=0"];
"test1.JPG" -- "test7.JPG"[label="Nm=0, Ni=0, C=0"];
}
```
I don't know how to solve this, maybe replace the 0 by a very small value when calling matchesGraphAsString?

**Comment by lukasalexanderweber** [comment]
Thanks for reproducing! I think it's an opencv internal issue. For know we should be fine throwing an error that `confidence must not be 0 if range matcher is used`. Do you want to implement it?

**Comment by Kalo0m** [comment]
Yes I can, I will try to do it this evening. But should I throw when the user wants to use graphAsString, or in the stitching constructor? Because everything seem to work fine except the string representation so would it be better to throw in the asString method?

**Comment by lukasalexanderweber** [comment]
Good question. 

> ```python
> 
> 
> stitcher.subset(imgs, features, matches) # works
> stitcher2.subset(imgs, features, matches) # doesn't work
> ```

so the stitching result is the same for both approaches? 

Then I would like your approach better: 

> maybe replace the 0 by a very small value when calling matchesGraphAsString

Maybe [here](https://github.com/lukasalexanderweber/stitching/blob/main/stitching/subsetter.py#L43) just ask if self.confidence_threshold is 0 and if yes, replace it with 0.0001 or so

**Comment by Kalo0m** [comment]
Yes if I remember well, only the graph representation doesn't work. The result is the same but I will double check this to be sure.

**Comment by Kalo0m** [comment]
@lukasalexanderweber done! https://github.com/lukasalexanderweber/stitching/pull/59

**Comment by lukasalexanderweber** [comment]
closed with dbc36e47eb7e963f07cc3297b848fae7f20b699a
